libs/wire-api/src/Wire/API/Routes/Cookies.hs

```diff
@@ -0,0 +1,139 @@
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Wire.API.Routes.Cookies where
+
+import Control.Error.Util
+import Data.List.NonEmpty (NonEmpty (..), nonEmpty)
+import qualified Data.Map as M
+import Data.Metrics.Servant
+import Data.SOP
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import GHC.TypeLits
+import Imports
+import Servant
+import Servant.Swagger
+import Web.Cookie (parseCookies)
+
+data (:::) a b
+
+data (::?) a b
+
+-- | A combinator to extract cookies from an HTTP request. The recommended way
+-- to use this combinator is to specify it exactly once in the description of
+-- an endpoint, passing a list of pairs of cookie name and type, separated by
+-- either '(:::)' or '(::?)'. The former makes the corresponding cookie
+-- mandatory, while the latter makes it optional, and returns a 'Maybe' result.
+--
+-- For example:
+-- @@
+-- Cookies '["foo" ::: Int64, "bar" ::? Text]
+-- @@
+-- results in a mandatory cookie with name "foo" containing a 64-bit integer,
+-- and an optional cookie with name "bar" containing an arbitrary text value.
+data Cookies (cs :: [*])
+
+type CookieHeader cs = Header' '[Required] "Cookie" (CookieTuple cs)
+
+-- CookieTypes = map snd
+type family CookieTypes (cs :: [*]) :: [*]
+
+type instance CookieTypes '[] = '[]
+
+type instance CookieTypes ((lbl ::: x) ': cs) = (NonEmpty x ': CookieTypes cs)
+
+type instance CookieTypes ((lbl ::? x) ': cs) = ([x] ': CookieTypes cs)
+
+newtype CookieTuple cs = CookieTuple {unCookieTuple :: NP I (CookieTypes cs)}
+
+type CookieMap = Map ByteString (NonEmpty ByteString)
+
+instance HasSwagger api => HasSwagger (Cookies cs :> api) where
+  toSwagger _ = toSwagger (Proxy @api)
+
+class CookieArgs (cs :: [*]) where
+  -- example: AddArgs ["foo" :: Foo, "bar" :: Bar] a = Foo -> Bar -> a
+  type AddArgs cs a :: *
+
+  uncurryArgs :: AddArgs cs a -> CookieTuple cs -> a
+  mapArgs :: (a -> b) -> AddArgs cs a -> AddArgs cs b
+  mkTuple :: CookieMap -> Either Text (CookieTuple cs)
+
+instance CookieArgs '[] where
+  type AddArgs '[] a = a
+  uncurryArgs a _ = a
+  mapArgs h = h
+  mkTuple _ = pure (CookieTuple Nil)
+
+instance
+  ( CookieArgs cs,
+    KnownSymbol lbl,
+    FromHttpApiData x
+  ) =>
+  CookieArgs ((lbl ::: (x :: *)) ': cs)
+  where
+  type AddArgs ((lbl ::: x) ': cs) a = NonEmpty x -> AddArgs cs a
+  uncurryArgs f (CookieTuple (I x :* xs)) = uncurryArgs @cs (f x) (CookieTuple xs)
+  mapArgs h f = mapArgs @cs h . f
+  mkTuple m = do
```

Comment:
Would they? Would you prefer something like the following?
```
uncurryArgs anArbitraryFunction (CookieTuple (I theFirstElementOfTheTuple :* theRestOfTheElementsOfTheTuple))
```

Honestly, this is type rejuggling. There is no substance. One letter names are fine, since the variables have no intrinsic meaning, they are just placeholders.

Reply:
Sorry I wasn't very clear with my comment, I had `lbl` and `cs` in mind. Also always a [fun read](https://degoes.net/articles/insufficiently-polymorphic).